spot-contracts/contracts/strategies/BasicFeeStrategy.sol

```diff
@@ -103,21 +119,46 @@ contract BasicFeeStrategy is IFeeStrategy, OwnableUpgradeable {
         emit UpdatedRolloverPerc(rolloverFeePerc_);
     }
 
+    /// @notice Enables debasement when the fee reserve is depleted.
+    function enableDebasement() public onlyOwner {
+        debasement = true;
+        emit UpdatedDebasementRule(true);
+    }
+
+    /// @notice Disables debasement when the fee reserve is depleted.
+    function disableDebasement() public onlyOwner {
+        debasement = false;
+        emit UpdatedDebasementRule(false);
+    }
+
     /// @inheritdoc IFeeStrategy
     function computeMintFees(uint256 mintAmt) external view override returns (int256, uint256) {
-        uint256 absoluteFee = (mintFeePerc.abs() * mintAmt) / HUNDRED_PERC;
-        return (mintFeePerc.sign() * absoluteFee.toInt256(), 0);
+        return (_computeFeeAmt(mintAmt, mintFeePerc), 0);
     }
 
     /// @inheritdoc IFeeStrategy
     function computeBurnFees(uint256 burnAmt) external view override returns (int256, uint256) {
-        uint256 absoluteFee = (burnFeePerc.abs() * burnAmt) / HUNDRED_PERC;
-        return (burnFeePerc.sign() * absoluteFee.toInt256(), 0);
+        return (_computeFeeAmt(burnAmt, burnFeePerc), 0);
     }
 
     /// @inheritdoc IFeeStrategy
     function computeRolloverFees(uint256 rolloverAmt) external view override returns (int256, uint256) {
-        uint256 absoluteFee = (rolloverFeePerc.abs() * rolloverAmt) / HUNDRED_PERC;
-        return (rolloverFeePerc.sign() * absoluteFee.toInt256(), 0);
+        return (_computeFeeAmt(rolloverAmt, rolloverFeePerc), 0);
+    }
+
+    /// @dev Given the token amount and fee percentage, computes the integer amount to be charged as fees.
+    function _computeFeeAmt(uint256 amount, int256 feePerc) private view returns (int256) {
+        uint256 absoluteFee = (feePerc.abs() * amount) / HUNDRED_PERC;
+        int256 feeAmt = feePerc.sign() * absoluteFee.toInt256();
+
+        // when fee is to be paid to the user or when debasement is enabled
+        // use the entire fee amount
+        if (feeAmt >= 0 || debasement) {
+            return feeAmt;
+        }
+
+        // fee is to be paid to the user and debasement is not-enabled
+        // pay out only till the reserve is depleted
+        return -1 * feeToken.balanceOf(feeReserve).min(absoluteFee).toInt256();
```

Comment:
This is a situation where `min(x, y)` makes more sense than using the inline "using" syntax.